File: autocontext/src/autocontext/server/protocol.py
"""WebSocket protocol models for the AutoContext TUI ↔ Server boundary.

This module is the single source of truth for the protocol. All message types
that flow over ``/ws/interactive`` are defined here as Pydantic models.

Use :func:`export_json_schema` to produce a JSON Schema document suitable for
cross-language validation (e.g. by the TypeScript TUI).
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter

PROTOCOL_VERSION = 1

# ---------------------------------------------------------------------------
# Nested / shared models
# ---------------------------------------------------------------------------


class ScenarioInfo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    description: str


class ExecutorResources(BaseModel):
    model_config = ConfigDict(extra="forbid")

    docker_image: str
    cpu_cores: int
    memory_gb: int
    disk_gb: int
    timeout_minutes: int


class ExecutorInfo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: str
    available: bool
    description: str
    resources: ExecutorResources | None = None


class StrategyParam(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    description: str


class ScoringComponent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    description: str
    weight: float


# ---------------------------------------------------------------------------
# Server -> Client messages
# ---------------------------------------------------------------------------


class HelloMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["hello"] = "hello"
    protocol_version: int = PROTOCOL_VERSION


class EventMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["event"] = "event"
    event: str
    payload: dict[str, Any]


class StateMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["state"] = "state"
    paused: bool
    generation: int = 0
    phase: str = ""


class ChatResponseMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["chat_response"] = "chat_response"
    role: str
    text: str


class EnvironmentsMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["environments"] = "environments"
    scenarios: list[ScenarioInfo]
    executors: list[ExecutorInfo]
    current_executor: str
    agent_provider: str


class RunAcceptedMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["run_accepted"] = "run_accepted"
    run_id: str
    scenario: str
    generations: int


class AckMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["ack"] = "ack"
    action: str
    decision: str | None = None


class ErrorMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["error"] = "error"
    message: str


class ScenarioGeneratingMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["scenario_generating"] = "scenario_generating"
    name: str


class ScenarioPreviewMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["scenario_preview"] = "scenario_preview"
    name: str
    display_name: str
    description: str
    strategy_params: list[StrategyParam]
    scoring_components: list[ScoringComponent]
    constraints: list[str]
    win_threshold: float


class ScenarioReadyMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["scenario_ready"] = "scenario_ready"
    name: str
    test_scores: list[float]


class ScenarioErrorMsg(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["scenario_error"] = "scenario_error"
    message: str
    stage: str


ServerMessage = Annotated[
    HelloMsg
    | EventMsg
    | StateMsg
    | ChatResponseMsg
    | EnvironmentsMsg
    | RunAcceptedMsg
    | AckMsg
    | ErrorMsg
    | ScenarioGeneratingMsg
    | ScenarioPreviewMsg
    | ScenarioReadyMsg
    | ScenarioErrorMsg,
    Field(discriminator="type"),
]


# ---------------------------------------------------------------------------
# Client -> Server messages
# ---------------------------------------------------------------------------


class PauseCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["pause"] = "pause"


class ResumeCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["resume"] = "resume"


class InjectHintCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["inject_hint"] = "inject_hint"
    text: str = Field(min_length=1)


class OverrideGateCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["override_gate"] = "override_gate"
    decision: Literal["advance", "retry", "rollback"]


class ChatAgentCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["chat_agent"] = "chat_agent"
    role: str
    message: str = Field(min_length=1)


class StartRunCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["start_run"] = "start_run"
    scenario: str
    generations: int = Field(gt=0)


class ListScenariosCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["list_scenarios"] = "list_scenarios"


class CreateScenarioCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["create_scenario"] = "create_scenario"
    description: str = Field(min_length=1)


class ConfirmScenarioCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["confirm_scenario"] = "confirm_scenario"


class ReviseScenarioCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["revise_scenario"] = "revise_scenario"
    feedback: str = Field(min_length=1)


class CancelScenarioCmd(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["cancel_scenario"] = "cancel_scenario"


ClientMessage = Annotated[
    PauseCmd
    | ResumeCmd
    | InjectHintCmd
    | OverrideGateCmd
    | ChatAgentCmd
    | StartRunCmd
    | ListScenariosCmd
    | CreateScenarioCmd
    | ConfirmScenarioCmd
    | ReviseScenarioCmd
    | CancelScenarioCmd,
    Field(discriminator="type"),
]


# ---------------------------------------------------------------------------
# Event payloads
# ---------------------------------------------------------------------------


class RunStartedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    scenario: str


class GenerationStartedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int


class AgentsStartedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    roles: list[str]


class RoleCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    role: str
    latency_ms: int
    tokens: int


class TournamentStartedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    matches: int


class MatchCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    match_index: int
    score: float


class TournamentCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    mean_score: float
    best_score: float
    wins: int
    losses: int


class GateDecidedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    decision: str
    delta: float


class CuratorStartedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int


class CuratorCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    decision: str


class GenerationCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    generation: int
    mean_score: float
    best_score: float
    elo: float
    gate_decision: str
    created_tools: list[str]


class RunCompletedPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    completed_generations: int


# ---------------------------------------------------------------------------
# Utility functions
# ---------------------------------------------------------------------------

_client_adapter: TypeAdapter[ClientMessage] = TypeAdapter(ClientMessage)


def parse_client_message(raw: dict[str, Any]) -> ClientMessage:
    """Validate and parse a raw dict into a typed client message.

    Raises ``ValidationError`` if the dict does not match any known message type.
    """
    return _client_adapter.validate_python(raw)


def export_json_schema() -> dict[str, Any]:
    """Export the full protocol as JSON Schema for cross-language validation."""
    return {
        "protocol_version": PROTOCOL_VERSION,
        "server_messages": TypeAdapter(ServerMessage).json_schema(),
        "client_messages": TypeAdapter(ClientMessage).json_schema(),
    }

File: ts/src/server/protocol.ts
/**
 * WebSocket protocol types — Zod schemas for client↔server messages (AC-347 Task 24).
 * Mirrors Python's autocontext/server/protocol.py.
 */

import { z } from "zod";

export const PROTOCOL_VERSION = 1;

// ---------------------------------------------------------------------------
// Nested models
// ---------------------------------------------------------------------------

export const ScenarioInfoSchema = z.object({
  name: z.string(),
  description: z.string(),
});

export const ExecutorResourcesSchema = z.object({
  docker_image: z.string(),
  cpu_cores: z.number(),
  memory_gb: z.number(),
  disk_gb: z.number(),
  timeout_minutes: z.number().int(),
});

export const ExecutorInfoSchema = z.object({
  mode: z.string(),
  available: z.boolean(),
  description: z.string(),
  resources: ExecutorResourcesSchema.optional().nullable(),
});

export const StrategyParamSchema = z.object({
  name: z.string(),
  description: z.string(),
});

export const ScoringComponentSchema = z.object({
  name: z.string(),
  description: z.string(),
  weight: z.number(),
});

// ---------------------------------------------------------------------------
// Server → Client messages
// ---------------------------------------------------------------------------

export const HelloMsgSchema = z.object({
  type: z.literal("hello"),
  protocol_version: z.number().int().optional(),
});

export const EventMsgSchema = z.object({
  type: z.literal("event"),
  event: z.string(),
  payload: z.record(z.unknown()),
});

export const StateMsgSchema = z.object({
  type: z.literal("state"),
  paused: z.boolean(),
  generation: z.number().int().optional(),
  phase: z.string().optional(),
});

export const ChatResponseMsgSchema = z.object({
  type: z.literal("chat_response"),
  role: z.string(),
  text: z.string(),
});

export const EnvironmentsMsgSchema = z.object({
  type: z.literal("environments"),
  scenarios: z.array(ScenarioInfoSchema),
  executors: z.array(ExecutorInfoSchema),
  current_executor: z.string(),
  agent_provider: z.string(),
});

export const RunAcceptedMsgSchema = z.object({
  type: z.literal("run_accepted"),
  run_id: z.string(),
  scenario: z.string(),
  generations: z.number().int(),
});

export const AckMsgSchema = z.object({
  type: z.literal("ack"),
  action: z.string(),
  decision: z.string().optional(),
});

export const ErrorMsgSchema = z.object({
  type: z.literal("error"),
  message: z.string(),
});

export const ScenarioGeneratingMsgSchema = z.object({
  type: z.literal("scenario_generating"),
  name: z.string(),
});

export const ScenarioPreviewMsgSchema = z.object({
  type: z.literal("scenario_preview"),
  name: z.string(),
  display_name: z.string(),
  description: z.string(),
  strategy_params: z.array(StrategyParamSchema),
  scoring_components: z.array(ScoringComponentSchema),
  constraints: z.array(z.string()),
  win_threshold: z.number(),
});

export const ScenarioReadyMsgSchema = z.object({
  type: z.literal("scenario_ready"),
  name: z.string(),
  test_scores: z.array(z.number()),
});

export const ScenarioErrorMsgSchema = z.object({
  type: z.literal("scenario_error"),
  message: z.string(),
  stage: z.string().optional(),
});

export const MonitorAlertMsgSchema = z.object({
  type: z.literal("monitor_alert"),
  alert_id: z.string(),
  condition_id: z.string(),
  condition_name: z.string(),
  condition_type: z.string(),
  scope: z.string(),
  detail: z.record(z.unknown()),
});

// Auth status response (AC-408)
export const AuthStatusMsgSchema = z.object({
  type: z.literal("auth_status"),
  provider: z.string(),
  authenticated: z.boolean(),
  model: z.string().optional(),
  configuredProviders: z.array(z.object({
    provider: z.string(),
    hasApiKey: z.boolean(),
  })).optional(),
});

// ---------------------------------------------------------------------------
// Client → Server commands
// ---------------------------------------------------------------------------

export const PauseCmdSchema = z.object({ type: z.literal("pause") });
export const ResumeCmdSchema = z.object({ type: z.literal("resume") });

export const InjectHintCmdSchema = z.object({
  type: z.literal("inject_hint"),
  text: z.string().min(1),
});

export const OverrideGateCmdSchema = z.object({
  type: z.literal("override_gate"),
  decision: z.enum(["advance", "retry", "rollback"]),
});

export const ChatAgentCmdSchema = z.object({
  type: z.literal("chat_agent"),
  role: z.string(),
  message: z.string(),
});

export const StartRunCmdSchema = z.object({
  type: z.literal("start_run"),
  scenario: z.string(),
  generations: z.number().int().positive(),
});

export const ListScenariosCmdSchema = z.object({
  type: z.literal("list_scenarios"),
});

export const CreateScenarioCmdSchema = z.object({
  type: z.literal("create_scenario"),
  description: z.string().min(1),
});

export const ConfirmScenarioCmdSchema = z.object({
  type: z.literal("confirm_scenario"),
});

export const ReviseScenarioCmdSchema = z.object({
  type: z.literal("revise_scenario"),
  feedback: z.string().min(1),
});

export const CancelScenarioCmdSchema = z.object({
  type: z.literal("cancel_scenario"),
});

// Auth commands (AC-408)
export const LoginCmdSchema = z.object({
  type: z.literal("login"),
  provider: z.string().min(1),
  apiKey: z.string().optional(),
  model: z.string().optional(),
  baseUrl: z.string().optional(),
});

export const LogoutCmdSchema = z.object({
  type: z.literal("logout"),
  provider: z.string().optional(),
});

export const SwitchProviderCmdSchema = z.object({
  type: z.literal("switch_provider"),
  provider: z.string().min(1),
});

export const WhoamiCmdSchema = z.object({
  type: z.literal("whoami"),
});

// ---------------------------------------------------------------------------
// Discriminated unions
// ---------------------------------------------------------------------------

export const ServerMessageSchema = z.discriminatedUnion("type", [
  HelloMsgSchema,
  EventMsgSchema,
  StateMsgSchema,
  ChatResponseMsgSchema,
  EnvironmentsMsgSchema,
  RunAcceptedMsgSchema,
  AckMsgSchema,
  ErrorMsgSchema,
  ScenarioGeneratingMsgSchema,
  ScenarioPreviewMsgSchema,
  ScenarioReadyMsgSchema,
  ScenarioErrorMsgSchema,
  MonitorAlertMsgSchema,
  AuthStatusMsgSchema,
]);

export const ClientMessageSchema = z.discriminatedUnion("type", [
  PauseCmdSchema,
  ResumeCmdSchema,
  InjectHintCmdSchema,
  OverrideGateCmdSchema,
  ChatAgentCmdSchema,
  StartRunCmdSchema,
  ListScenariosCmdSchema,
  CreateScenarioCmdSchema,
  ConfirmScenarioCmdSchema,
  ReviseScenarioCmdSchema,
  CancelScenarioCmdSchema,
  LoginCmdSchema,
  LogoutCmdSchema,
  SwitchProviderCmdSchema,
  WhoamiCmdSchema,
]);

export type ServerMessage = z.infer<typeof ServerMessageSchema>;
export type ClientMessage = z.infer<typeof ClientMessageSchema>;

export function parseClientMessage(raw: Record<string, unknown>): ClientMessage {
  return ClientMessageSchema.parse(raw);
}

export function parseServerMessage(raw: Record<string, unknown>): ServerMessage {
  return ServerMessageSchema.parse(raw);
}
